templates/bucket-replication-lambda.yaml

```diff
@@ -0,0 +1,200 @@
+      See https://docs.synapse.org/articles/custom_storage_location.html
+  SynapseNotifyUsers:
+    Type: String
+    Description: >
+      Comma-delimited list of numeric Synapse user IDs to be notified when an attempt to sync to Synapse is made
+      by the Lambda function. Example: the `ownerId` returned by a call to the Synapse REST API /userProfile.
+  ReplicationSourceAccountId:
+    Type: Number
+    Description: ID of the AWS account holding the replication source bucket.
+  SynapseAWSAccountId:
+    Type: Number
+    Description: >
+      ID of the Synapse AWS account. Used for allowing Synapse to access the DataBucket as an external storage location.
+      See https://docs.synapse.org/articles/custom_storage_location.html
+    Default: 325565585839
+  Md5BlockSize:
+    Type: Number
+    Default: 5242880 # 5 MB
+    Description:
+      Maximum block size allocated to computing MD5 sums, in bytes.
+
+Resources:
+
+  LambdaExecutionRole:
+    Type: AWS::IAM::Role
+    Properties:
+      #Description: ARN of role that lambda uses for execution.
+      RoleName: !Sub ${AWS::StackName}-LambdaExecutionRole
+      AssumeRolePolicyDocument:
+        Statement:
+          - Action: ['sts:AssumeRole']
+            Effect: Allow
+            Principal:
+              Service: [lambda.amazonaws.com]
+
+  LambdaExecutionRolePolicy:
+    Type: AWS::IAM::Policy
+    Properties:
+      #Description: Allows access to the S3 data bucket and CloudWatch logs.
+      PolicyName: !Sub ${AWS::StackName}-LambdaExecutionPolicy
+      Roles: [!Ref 'LambdaExecutionRole']
+      PolicyDocument:
+        Statement:
+          - Effect: Allow
+            Action:
+              - logs:CreateLogGroup
+              - logs:CreateLogStream
+              - logs:PutLogEvents
+            Resource: "*"
+          - Effect: Allow
+            Action:
+              - s3:ListBucket
+            Resource:
+              - !Join
+                - ""
+                - - "arn:aws:s3:::"
+                  - !Sub ${AWS::StackName}-data
+            Sid: VisualEditor1
+          - Effect: Allow
+            Action:
+              - s3:PutObject
+              - s3:GetObject
+            Resource:
+              - !Join
+                - ""
+                - - "arn:aws:s3:::"
+                  - !Sub ${AWS::StackName}-data
+                  - "/*"
+
+  BucketToSynapseFunction:
+    Type: AWS::Lambda::Function
+    Properties:
+      FunctionName: !Ref FunctionName
+      Code:
+        S3Bucket: !Ref FunctionCodeBucket
+        S3Key: !Ref FunctionCodeKey
+      Description: Adds files from an S3 bucket to Synapse.
+      Environment:
+        Variables:
+          SYNAPSE_USERNAME: !Ref SynapseUsername
+          SYNAPSE_APIKEY: !Ref SynapseApikey
+          SYNAPSE_SOURCE_BUCKET_STORAGE_LOCATION_ID: !Ref SynapseStorageLocationId
+          SYNAPSE_NOTIFY_USERS: !Ref SynapseNotifyUsers
+          MD5_BLOCK_SIZE: !Ref Md5BlockSize
```

Comment:
don't you want to pass in the `DataBucket` (destination bucket) to know where to write the data to?

Reply:
Thanks for checking on that! It might be klunky but the Lambda code assumes there is an [external storage location](https://docs.synapse.org/articles/custom_storage_location.html) set up in Synapse, whose id is SynapseStorageLocationID that points to the DataBucket. So it does not need to reference the DataBucket directly.

That does mean that after the CF template builds the stack, someone needs to go in to Synapse by hand (or script) and create the storage location.

The Lambda can't do that itself because AFAIK there's no way to check whether an external storage location has already been created in Synapse. And the Lambda function will execute many times and I do not want to create dups.

Reply:
Oh, and you are probably wondering how the Lambda knows to look in the DataBucket for incoming files too...  when the Lambda is triggered, it passes an `event` argument with the bucket and key of the S3 object that triggered it, to the lambda handler code.

Reply:
Good to know. It won't work in this case because my bucket needs a different BucketPolicy than the one in that template.